<a href="https://colab.research.google.com/github/jackson51472/RNAKeras/blob/main/RNAKeras.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [50]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.metrics import accuracy_score, f1_score, confusion_matrix, recall_score
from keras.models import Sequential
from keras.layers import Dense, Dropout


In [51]:
df = pd.read_csv('Churn_treino.csv', sep=';')
df

CreditScore Geography  Gender  Age  Tenure   Balance  NumOfProducts  \
0             619    France  Female   42       2         0              1   
1             608     Spain  Female   41       1   8380786              1   
2             502    France  Female   42       8   1596608              3   
3             699    France  Female   39       1         0              2   
4             850     Spain  Female   43       2  12551082              1   
...           ...       ...     ...  ...     ...       ...            ...   
9995          771    France    Male   39       5         0              2   
9996          516    France    Male   35      10   5736961              1   
9997          709    France  Female   36       7         0              1   
9998          772   Germany    Male   42       3   7507531              2   
9999          792    France  Female   28       4  13014279              1   

      HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0             1               1         10134888       1  
1             0               1         11254258       0  
2             1               0         11393157       1  
3             0               0          9382663       0  
4             1               1           790841       0  
...         ...             ...              ...     ...  
9995          1               0          9627064       0  
9996          1               1         10169977       0  
9997          0               1          4208558       1  
9998          1               0          9288852       1  
9999          1               0          3819078       0  

[10000 rows x 11 columns]

In [52]:
X = df.drop('Exited', axis=1)
y = df['Exited']

In [53]:
# Fazendo a padronização das colunas
# Apenas aonde não for string
stand = StandardScaler()
numerico = X.select_dtypes(include=['int64', 'float64']).columns
X[numerico] = stand.fit_transform(X[numerico])



In [54]:
#Modificar as colunas Geography e Gender para numeros.
lbEncoder= LabelEncoder()
categorico = X.select_dtypes(include=['object']).columns
X[categorico] = X[categorico].apply(lbEncoder.fit_transform)
X

CreditScore  Geography  Gender       Age    Tenure   Balance  \
0       -0.326221          0       0  0.293517 -1.041760 -1.110553   
1       -0.440036          2       0  0.198164 -1.387538  0.222782   
2       -1.536794          0       0  0.293517  1.032908 -0.856542   
3        0.501521          0       0  0.007457 -1.387538 -1.110553   
4        2.063884          2       0  0.388871 -1.041760  0.886252   
...           ...        ...     ...       ...       ...       ...   
9995     1.246488          0       1  0.007457 -0.004426 -1.110553   
9996    -1.391939          0       1 -0.373958  1.724464 -0.197835   
9997     0.604988          0       0 -0.278604  0.687130 -1.110553   
9998     1.256835          1       1  0.293517 -0.695982  0.083852   
9999     1.463771          0       0 -1.041433 -0.350204  0.959944   

      NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  
0         -0.911583   0.646092        0.970243         0.170614  
1         -0.911583  -1.547768        0.970243         0.353281  
2          2.527057   0.646092       -1.030670         0.375948  
3          0.807737  -1.547768       -1.030670         0.047859  
4         -0.911583   0.646092        0.970243        -1.354223  
...             ...        ...             ...              ...  
9995       0.807737   0.646092       -1.030670         0.087743  
9996      -0.911583   0.646092        0.970243         0.176340  
9997      -0.911583  -1.547768        0.970243        -0.796492  
9998       0.807737   0.646092       -1.030670         0.032551  
9999      -0.911583   0.646092       -1.030670        -0.860051  

[10000 rows x 10 columns]

In [55]:
x_treinamento, x_teste, y_treinamento, y_teste = train_test_split(X, y, test_size=0.2, random_state=0)


In [56]:
# Montando o formado da rede
modelo = Sequential()
modelo.add(Dense(units=120, activation='relu', input_dim=x_treinamento.shape[1]))
modelo.add(Dropout(0.2))
modelo.add(Dense(units=120, activation='relu'))
modelo.add(Dropout(0.3))
modelo.add(Dense(units=120, activation='relu'))
modelo.add(Dropout(0.4))
modelo.add(Dense(units=1, activation='sigmoid'))

/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [57]:
modelo.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
modelo.fit(x_treinamento, y_treinamento, epochs=100, batch_size=32)

Epoch 1/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7806 - loss: 0.4972
Epoch 2/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8383 - loss: 0.3964
Epoch 3/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8464 - loss: 0.3716
Epoch 4/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8454 - loss: 0.3708
Epoch 5/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8551 - loss: 0.3624
Epoch 6/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8525 - loss: 0.3649
Epoch 7/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8588 - loss: 0.3527
Epoch 8/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8476 - loss: 0.3618
Epoch 9/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.8589 - loss: 0.3433
Epoch 10/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.8572 - loss: 0.3512
Epoch 11/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8543 - loss: 0.3591
Epoch 12/100
250/250 ━━━━━━━━━━━━━━━━━━━━

In [58]:
previsao = modelo.predict(x_teste)
y_pred = (previsao > 0.5).astype('int32')
y_pred


63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step


array([[0],
       [0],
       [0],
       ...,
       [0],
       [0],
       [0]], dtype=int32)

In [59]:
print('Acurácia: ', accuracy_score(y_teste, y_pred))
print('F1: ', f1_score(y_teste, y_pred))
print('Recall: ', recall_score(y_teste, y_pred))
print('Matriz de Confusão: ', confusion_matrix(y_teste, y_pred))

Acurácia:  0.8495
F1:  0.5553914327917282
Recall:  0.4641975308641975
Matriz de Confusão:  [[1511   84]
 [ 217  188]]


# TESTE 1
### 64 32 64 com 100 epoch
**Acurácia**:  0.862

**F1**:  0.5843373493975904

**Recall**:  0.47901234567901235

**Matriz de Confusão**:  [[1530   65]
 [ 211  194]]


---
# TESTE 2
### 64 32 16 com 100 epoch

**Acurácia**:  0.8615

**F1**:  0.5718701700154559

**Recall**:  0.4567901234567901

**Matriz de Confusão**:  [[1538   57]
 [ 220  185]]



---
 # TESTE 3
### 32 64 32 com 100 epoch
**Acurácia**:  0.863

**F1**:  0.5861027190332326

**Recall**:  0.47901234567901235

**Matriz de Confusão**:  [[1532   63]
 [ 211  194]]

 ---
 # TESTE 4
 ### 120 60 80 com 100 epoch

 **Acurácia**:  0.853

**F1**:  0.5823863636363636

**Recall**:  0.5061728395061729

**Matriz de Confusão**:  [[1501   94]
 [ 200  205]]

 ---
 # TESTE 5
 ### 120 60 80 com 100 epoch

 **Acurácia**:  0.8475

**F1**:  0.5547445255474452

**Recall**:  0.4691358024691358

**Matriz de Confusão**:  [[1505   90]
 [ 215  190]]

---

#Opinião
